types/src/transaction/analyzed_transaction.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access_path::AccessPath,
+    account_config::{AccountResource, CoinStoreResource},
+    state_store::{state_key::StateKey, table::TableHandle},
+    transaction::{SignedTransaction, Transaction, TransactionPayload},
+};
+use aptos_crypto::{
+    hash::{CryptoHash, DummyHasher},
+    HashValue,
+};
+pub use move_core_types::abi::{
+    ArgumentABI, ScriptFunctionABI as EntryFunctionABI, TransactionScriptABI, TypeArgumentABI,
+};
+use move_core_types::{
+    account_address::AccountAddress, language_storage::StructTag, move_resource::MoveStructType,
+};
+use std::hash::{Hash, Hasher};
+
+#[derive(Clone, Debug)]
+pub struct AnalyzedTransaction {
+    transaction: Transaction,
+    /// Set of storage locations that are read by the transaction - this doesn't include location
+    /// that are written by the transactions to avoid duplication of locations across read and write sets
+    /// This can be accurate or strictly overestimated.
+    read_set: Vec<StorageLocation>,
+    /// Set of storage locations that are written by the transaction. This can be accurate or strictly
+    /// overestimated.
+    write_set: Vec<StorageLocation>,
+    /// A transaction is predictable if neither the read_hint or the write_hint have wildcards.
+    predictable_transaction: bool,
+    /// The hash of the transaction - this is cached for performance reasons.
+    hash: HashValue,
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum StorageLocation {
+    // A specific storage location denoted by an address and a struct tag.
+    Specific(StateKey),
+    // Storage location denoted by a struct tag and any arbitrary address.
+    // Example read<T>(*), write<T>(*) in Move
+    WildCardStruct(StructTag),
+    // Storage location denoted by a table handle and any arbitrary item in the table.
+    WildCardTable(TableHandle),
+}
+
+impl CryptoHash for StorageLocation {
+    type Hasher = DummyHasher;
+
+    fn hash(&self) -> HashValue {
+        match self {
+            StorageLocation::Specific(state_key) => CryptoHash::hash(state_key),
+            _ => todo!("hashing of wildcard storage location is not supported yet"),
```

Comment:
For other variants as well, we will need to evaluate. In the worst case, we can cache the hash in the storage location itself, if the variants themselves don't cache. But avoid the hash calculation is critical for performance of the partitioning. LMK your thoughts.

Reply:
well, it'll be a domain separation issue if we don't distinguish between variants IIUC. @alinush @zjma

Reply:
Can you clarify why there might be conflict if we don't have domain separation? In any case, once we introduce other variants, we can have variants cache the hash or we can change this to cache the hash in the storage location itself. For now, its not needed as the hash is already cached in `StateKey`.

Reply:
I mean the hash of a StateKey (which is also that of `StorageLocation::Specifict(key)` of the same key) can conflict with that of a `StorageLocation::WildCard(type)`

Reply:
actually why does `StorageLocation` need to implement `CryptoHash`? IIRC it only needs to be the key of some hash map.

Reply:
> I mean the hash of a StateKey (which is also that of `StorageLocation::Specifict(key)` of the same key) can conflict with that of a `StorageLocation::WildCard(type)`

@msmouse - In that case, we can cache the hash in the `StorageLocation` itself when we add the new variants? For now, we don't need it as we are supporting only one. 


@zjma - Really `CryptoHash` is not needed but Since we already cache the `CryptoHash` for underlying `StateKey`, we can just reuse it for free without hash recalculation.

Reply:
@zjma  Good point.. 
@sitalkedia  Looks like you are not getting the benefit because the hash sets uses the derived `Hash` trait..

Reply:
should be:
```
impl Hash for StorageLocation {
    fn hash<H: Hasher>(&self, state: &mut H) {
        match &self {
            Specific(key) => CryptoHash::hash(key),
            ...
        }
    }
}
```
i guess

Reply:
And let's note in the code that this is not conflict free, so can't be simply changed to CryptoHash implementation directly. -- just in case. (sorry I'm a little paranoid about it since evil @alinush sit close to me)

Reply:
Good point @msmouse , @zjma . I have removed the `CryptoHash` for now - we were using this in one place but that didn't cause any noticeable regression. 

Regarding conflict, that should not matter as we are using this only for hashmap lookup i.e., we don't need conflict free semantic like the `StateKey`